packages/core/src/subscriptions/index.ts

```diff
@@ -0,0 +1,259 @@
+import { Bundle, Parameters, Subscription, SubscriptionStatus } from '@medplum/fhirtypes';
+import { MedplumClient } from '../client';
+import { TypedEventTarget } from '../eventtarget';
+import { OperationOutcomeError, badRequest, serverError, validationError } from '../outcomes';
+import { ProfileResource, getReferenceString, resolveId } from '../utils';
+
+export type SubscriptionEventMap = {
+  connect: { type: 'connect'; payload: { subscriptionId: string } };
+  disconnect: { type: 'disconnect'; payload: { subscriptionId: string } };
+  error: { type: 'error'; payload: Error };
+  message: { type: 'message'; payload: Bundle };
+  close: { type: 'close' };
+  heartbeat: { type: 'heartbeat'; payload: Bundle };
+};
+
+const kAddCriteria = Symbol.for('medplum.SubscriptionEmitter.addCriteria');
+const kRemoveCriteria = Symbol.for('medplum.SubscriptionEmitter.removeCriteria');
+
+/**
+ * An `EventTarget` that emits events when new subscription notifications come in over WebSockets.
+ *
+ * -----
+ *
+ * ### Events emitted:
+ *
+ * - `connect` - A new subscription is connected to the `SubscriptionManager` and `message` events for this subscription can be expected.
+ * - `disconnect` - The specified subscription is no longer being monitored by the `SubscriptionManager`.
+ * - `error` - An error has occurred.
+ * - `message` - A message containing a notification `Bundle` has been received.
+ * - `close` - The WebSocket has been closed.
+ * - `heartbeat` - A `heartbeat` message has been received.
+ */
+export class SubscriptionEmitter extends TypedEventTarget<SubscriptionEventMap> {
+  private criteria: Set<string>;
+  constructor(...criteria: string[]) {
+    super();
+    this.criteria = new Set(criteria);
+  }
+  getCriteria(): Set<string> {
+    return this.criteria;
+  }
+  [kAddCriteria](criteria: string): void {
+    this.criteria.add(criteria);
+  }
+  [kRemoveCriteria](criteria: string): void {
+    this.criteria.delete(criteria);
+  }
+}
+
+export class SubscriptionManager {
+  private readonly medplum: MedplumClient;
+  private ws: WebSocket;
+  private masterSubEmitter?: SubscriptionEmitter;
+  private subEmitters: Map<string, SubscriptionEmitter>; // Map<criteria, SubscriptionEmitter>
+  private refCounts: Map<string, number>; // Map<criteria, refCount>
+  private subscriptionCriteriaLookup: Map<string, string>; // Map<subscriptionId, criteria>
+  private criteriaSubscriptionLookup: Map<string, string>; // Map<criteria, subscriptionId>
+  private wsClosed: boolean;
+
+  constructor(medplum: MedplumClient, wsOrUrl: WebSocket | string) {
```

Comment:
Looking at all of the call sites for `new SubscriptionManager`, and specifically cases where the 2nd arg is a `WebSocket`...

Is the `WebSocket` option ever required?  It looks like this could always take a `string`, and then you could simplify the constructor and eliminate a bunch of tests.

Reply:
Yes sometimes you need the caller to have a handle on the `WebSocket`, particularly for testing purposes, but I could see that you might want that in general as well.

Reply:
Actually looks like I might have removed the tests where I was doing that, good call. We can probably simplify that.

Reply:
Now we accept `URL` or `string` here, and simplified the code 👍